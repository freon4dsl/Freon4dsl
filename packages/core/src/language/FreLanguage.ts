import { ReferenceShortcut } from "../editor";
import { FreNode, FreModel, FreModelUnit } from "../ast";
import { EmptyStdLib, FreStdlib } from "../stdlib/index";
import { isNullOrUndefined } from "../util";
// import { FreLogger } from "../logging";
// const LOGGER = new FreLogger("Language");

export type PropertyKind = "primitive" | "part" | "reference";

export type PrimType = string | boolean | number;

export type FreLanguageProperty = {
    name: string;
    type: string;
    id?: string;
    key?: string; // used for LionWeb
    isList: boolean;
    isPublic: boolean;
    language: string;
    propertyKind: PropertyKind;
};
export type FreLanguageModel = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    language: string;
    subConceptNames?: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreModel;
    creator: (data: Partial<FreModel>) => FreModel;
    referenceShortcut?: ReferenceShortcut;
};
export type FreLanguageModelUnit = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    // isPublic?: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    language: string;
    subConceptNames: string[];
    fileExtension: string;
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreModelUnit;
    creator: (data: Partial<FreModelUnit>) => FreModelUnit;
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};
export type FreLanguageConcept = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isAbstract: boolean;
    isPublic: boolean;
    isLimited: boolean;
    instanceNames: string[];
    isNamespace?: boolean;
    isNamedElement?: boolean;
    language: string;
    baseName: string;
    subConceptNames: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor: (id?: string) => FreNode;
    creator: (data: Partial<FreNode>) => FreNode;
    // Used by editor, therefore only in Concept
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};

export type FreLanguageInterface = {
    typeName: string;
    id?: string;
    key?: string; // used for LionWeb
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    isAbstract?: boolean;
    subConceptNames: string[];
    properties: Map<string, FreLanguageProperty>;
    constructor?: (id?: string) => FreNode | undefined;
    creator?: (data: Partial<FreNode>) => FreNode | undefined;
    language: string;
    referenceShortcut?: ReferenceShortcut;
};

export type FreLanguageClassifier = FreLanguageModel | FreLanguageModelUnit | FreLanguageConcept | FreLanguageInterface;

export class FreLanguage {
    private static theInstance: FreLanguage;

    static getInstance() {
        if (FreLanguage.theInstance === undefined) {
            FreLanguage.theInstance = new FreLanguage();
        }
        return FreLanguage.theInstance;
    }

    private languageName: string;
    private languageId?: string;
    private pmodel: FreLanguageModel;
    private units: Map<string, FreLanguageModelUnit> = new Map<string, FreLanguageModelUnit>();
    private concepts: Map<string, FreLanguageConcept> = new Map<string, FreLanguageConcept>();
    private interfaces: Map<string, FreLanguageInterface> = new Map<string, FreLanguageInterface>();
    private _stdLib: FreStdlib = new EmptyStdLib();
    set stdLib(lib: FreStdlib) {
        this._stdLib = lib;
    }
    get stdLib(): FreStdlib {
        return this._stdLib;
    }

    private constructor() {}

    model(): FreLanguageModel {
        return this.pmodel;
    }

    modelOfType(typeName: string) {
        if (!!this.pmodel && this.pmodel.typeName === typeName) {
            return this.pmodel;
        } else {
            return null;
        }
    }

    unit(typeName: string): FreLanguageModelUnit | undefined {
        return this.units.get(typeName);
    }

    unitByKey(key: string): FreLanguageModelUnit | undefined {
        return this.helperByKey(this.units, key) as FreLanguageModelUnit;
    }

    concept(typeName: string): FreLanguageConcept | undefined {
        // console.log("Language find concept " + typeName);
        return this.concepts.get(typeName);
    }

    conceptByKey(conceptKey: string): FreLanguageConcept | undefined {
        return this.helperByKey(this.concepts, conceptKey) as FreLanguageConcept;
    }

    private helperByKey(
        map: Map<string, FreLanguageClassifier>,
        conceptKey: string,
    ): FreLanguageClassifier | undefined {
        for (const concept of map.values()) {
            if (concept.key === conceptKey) {
                return concept;
            }
        }
        return null;
    }

    interface(typeName: string): FreLanguageInterface | undefined {
        return this.interfaces.get(typeName);
    }

    interfaceByKey(key: string): FreLanguageInterface | undefined {
        return this.helperByKey(this.interfaces, key) as FreLanguageInterface;
    }

    classifier(typeName: string): FreLanguageClassifier | undefined {
        const concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1;
        } else {
            const intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf;
            } else {
                const unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1;
                } else {
                    const model = this.modelOfType(typeName);
                    if (!!model) {
                        return model;
                    }
                }
            }
        }
        // console.log("RETURNING NULL FOR " + typeName)
        return undefined;
    }

    classifierByKey(key: string): FreLanguageClassifier | undefined {
        const concept1 = this.conceptByKey(key);
        if (!!concept1) {
            return concept1;
        } else {
            const intf = this.interfaceByKey(key);
            if (!!intf) {
                return intf;
            } else {
                const unit1 = this.unitByKey(key);
                if (!!unit1) {
                    return unit1;
                } else {
                    // TODO By Id for models
                    const model = this.modelOfType(key);
                    if (!!model) {
                        return model;
                    }
                }
            }
        }
        // console.log("RETURNING NULL FOR " + typeName)
        return undefined;
    }

    // private conceptProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.concepts.get(typeName)?.properties.get(propertyName);
    // }
    //
    // private conceptPropertyByKey(conceptKey: string, propertyKey: string): FreLanguageProperty | undefined {
    //     return this.helperPropByKey(this.conceptByKey(conceptKey).properties, propertyKey);
    // }

    helperPropByKey(map: Map<string, FreLanguageProperty>, key: string): FreLanguageProperty | undefined {
        for (const prop of map.values()) {
            if (prop.key === key) {
                return prop;
            }
        }
        return undefined;
    }

    // private unitProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.units.get(typeName)?.properties.get(propertyName);
    // }

    unitPropertyByKey(unitKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.helperPropByKey(this.unitByKey(unitKey).properties, propertyKey);
    }

    // private interfaceProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
    //     return this.interfaces.get(typeName)?.properties.get(propertyName);
    // }

    interfacePropertyByKey(interfaceKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.helperPropByKey(this.interfaceByKey(interfaceKey).properties, propertyKey);
    }

    classifierProperty(typeName: string, propertyName: string): FreLanguageProperty | undefined {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        const concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1.properties.get(propertyName);
        } else {
            const intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf.properties.get(propertyName);
            } else {
                const unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1.properties.get(propertyName);
                } else {
                    const model = this.modelOfType(typeName);
                    if (!!model) {
                        return model.properties.get(propertyName);
                    }
                }
            }
        }
        return undefined;
    }

    classifierPropertyByKey(classifierKey: string, propertyKey: string): FreLanguageProperty | undefined {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        const concept1 = this.classifierByKey(classifierKey);
        return this.helperPropByKey(concept1.properties, propertyKey);
    }

    allConceptProperties(typeName: string): FreLanguageProperty[] | undefined {
        // console.log("Looking up properties for "+ typeName);
        let myType: FreLanguageConcept | FreLanguageModelUnit | undefined = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        if (myType === undefined) {
            return [];
        }
        return [...myType.properties.values()];
    }

    /**
     * Get all properties of kind "kind" of classifier "typename"
     * @param typename
     * @param ptype
     */
    public getPropertiesOfKind(typename: string, ptype: PropertyKind): FreLanguageProperty[] {
        const classifier: FreLanguageClassifier | undefined = FreLanguage.getInstance().classifier(typename);
        const foundProperties: FreLanguageProperty[] = [];
        if (!!classifier) {
            for (const prop of classifier.properties.values()) {
                if (prop.propertyKind === ptype) {
                    foundProperties.push(prop);
                }
            }
        }
        return foundProperties;
    }

    /**
     * Return the value of `prop' in `element'.
     * For ease of use, always returns a list, even is the property is not a list.
     * @param element
     * @param prop
     */
    public getPropertyValue(element: FreNode, prop: FreLanguageProperty): FreNode[] {
        if (prop.isList) {
            return element[prop.name];
        } else {
            const value = element[prop.name];
            if (!!value) {
                return [value];
            } else {
                return [];
            }
        }
    }

    /**
     * Return all named concept in the language.
     */
    public getNamedConcepts(): string[] {
        return Array.from(this.concepts.values())
            .filter((concept) => concept.isNamedElement)
            .map((concept) => concept.typeName);
    }

    /**
     * Return all named interfaces in the language.
     */
    public getNamedInterfaces(): string[] {
        return Array.from(this.interfaces.values())
            .filter((intfc) => intfc.isNamedElement)
            .map((intfc) => intfc.typeName);
    }

    /**
     * Return all named concepts and interfaces in the language.
     */
    public getNamedElements(): string[] {
        return this.getNamedConcepts().concat(this.getNamedInterfaces());
    }

    /**
     * Return the names of all model unit types.
     */
    public getUnitNames(): string[] {
        return Array.from(this.units.values()).map((unit) => unit.typeName);
    }

    createModel(id?: string): FreModel {
        return this.pmodel?.constructor(id);
    }

    createUnit(typeName: string, id?: string): FreModelUnit | undefined {
        return this.units.get(typeName)?.constructor(id);
    }

    /**
     * Create a new instance of the class `typeName`.
     * @param typeName
     */
    createConceptOrUnit(typeName: string, id?: string): FreNode | undefined {
        let myType: FreLanguageConcept | FreLanguageModelUnit | undefined = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.constructor(id);
    }

    addModel(model: FreLanguageModel) {
        if (!!this.pmodel) {
            console.error(
                "Language: adding model of type " +
                    model?.typeName +
                    " while there is already a model of type " +
                    this.pmodel.typeName,
            );
        }
        this.pmodel = model;
    }

    addUnit(unit: FreLanguageModelUnit) {
        this.units.set(unit.typeName, unit);
        unit.subConceptNames = [];
    }

    /**
     * Add a concept definition to this language
     * @param concept
     */
    addConcept(concept: FreLanguageConcept) {
        this.concepts.set(concept.typeName, concept);
    }

    addInterface(intface: FreLanguageInterface) {
        this.interfaces.set(intface.typeName, intface);
    }

    /**
     * Set Language name
     * @param name
     */
    set name(name: string) {
        this.languageName = name;
    }

    get name(): string {
        return this.languageName;
    }

    set id(id: string) {
        this.languageId = id;
    }

    get id(): string {
        return this.languageId;
    }

    subConcepts(typeName: string): string[] {
        const concept = this.concept(typeName);
        if (!!concept) {
            return concept.subConceptNames;
        }
        const intface = this.interface(typeName);
        if (!!intface) {
            return intface.subConceptNames;
        }
        return [];
    }

    referenceCreator: (name: string | string[], type: string) => any;

    addReferenceCreator(creator: (name: string | string[], type: string) => any) {
        this.referenceCreator = creator;
    }

    /**
     * Returns true if the freLanguageConcept of 'element', i.e. its metatype,
     * is the same as 'requestedType' or is a subtype of 'requestedType'.
     * @param element
     * @param requestedType
     */
    public metaConformsToType(element: FreNode, requestedType: string): boolean {
        const metatype = element.freLanguageConcept();
        return metatype === requestedType || FreLanguage.getInstance().subConcepts(requestedType).includes(metatype);
    }
}
