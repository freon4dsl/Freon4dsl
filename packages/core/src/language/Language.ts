import { ReferenceShortcut } from "../editor";
import { PiElement, PiModel, PiModelUnit } from "../ast";
import { isNullOrUndefined } from "../util";
import { PiLogger } from "../logging";
const LOGGER = new PiLogger("Language");

export type PropertyKind = "primitive" | "part" | "reference";

export type Property = {
    name: string;
    type: string;
    isList: boolean;
    isPublic: boolean;
    propertyKind: PropertyKind;
};
export type Model = {
    typeName: string;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames?: string[];
    properties: Map<string, Property>;
    constructor: () => PiModel;
};
export type ModelUnit = {
    typeName: string;
    // isPublic?: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames?: string[];
    fileExtension: string;
    properties: Map<string, Property>;
    constructor: () => PiModelUnit;
};
export type Concept = {
    typeName: string;
    isAbstract: boolean;
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    baseName: string;
    subConceptNames: string[];
    properties: Map<string, Property>;
    constructor: () => PiElement;
    // Used by editor, therefore only in Concept
    trigger: string;
    referenceShortcut?: ReferenceShortcut;
};

export type Interface = {
    typeName: string;
    isPublic: boolean;
    isNamespace?: boolean;
    isNamedElement?: boolean;
    subConceptNames: string[];
    properties: Map<string, Property>;
};

export type Classifier = Model | ModelUnit | Concept | Interface;

export class Language {
    private static theInstance: Language = null;

    static getInstance() {
        if (Language.theInstance === null) {
            Language.theInstance = new Language();
        }
        return Language.theInstance;
    }

    private languageName: string;
    private pmodel: Model;
    private units: Map<string, ModelUnit> = new Map<string, ModelUnit>();
    private concepts: Map<string, Concept> = new Map<string, Concept>();
    private interfaces: Map<string, Interface> = new Map<string, Interface>();

    private constructor() {
    }

    model(): Model {
        return this.pmodel;
    }

    modelOfType(typeName: string) {
        if(!!this.pmodel && (this.pmodel.typeName === typeName)) {
            return this.pmodel;
        } else {
            return null;
        }
    }

    unit(typeName: string): ModelUnit {
        return this.units.get(typeName);
    }

    concept(typeName: string): Concept {
        // console.log("Language find concept " + typeName);
        return this.concepts.get(typeName);
    }

    interface(typeName: string): Interface {
        return this.interfaces.get(typeName);
    }

    classifier(typeName: string): Classifier {
        let concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1;
        } else {
            let intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf;
            } else {
                let unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1;
                } else {
                    let model = this.modelOfType(typeName);
                    if(!! model){
                        return model;
                    }
                }
            }
        }
        console.log("RETURNING NULL FOR " + typeName)
        return null;
    }

    conceptProperty(typeName: string, propertyName: string): Property {
        // LOGGER.log("copnceptProperty [" + typeName + "."  + propertyName + "]");
        return this.concepts.get(typeName).properties.get(propertyName);
    }

    unitProperty(typeName: string, propertyName: string): Property {
        return this.units.get(typeName).properties.get(propertyName);
    }

    interfaceProperty(typeName: string, propertyName: string): Property {
        return this.interfaces.get(typeName).properties.get(propertyName);
    }

    classifierProperty(typeName: string, propertyName: string): Property {
        // LOGGER.log("CLASSIFIERPROPERTY " + typeName + "." + propertyName);
        let concept1 = this.concepts.get(typeName);
        if (!!concept1) {
            return concept1.properties.get(propertyName);
        } else {
            let intf = this.interfaces.get(typeName);
            if (!!intf) {
                return intf.properties.get(propertyName);
            } else {
                let unit1 = this.units.get(typeName);
                if (!!unit1) {
                    return unit1.properties.get(propertyName);
                }  else {
                    let model = this.modelOfType(typeName);
                    if( !!model){
                        return model.properties.get(propertyName);
                    }
                }
            }
        }
        return null;
    }

    allConceptProperties(typeName: string): IterableIterator<Property> {
        // console.log("Looking up properties for "+ typeName);
        let myType: Concept | ModelUnit = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.properties.values();
    }

    /**
     * Get all properties of kind "kind" of classifier "typename"
     * @param typename
     * @param ptype
     */
    public getPropertiesOfKind(typename: string, ptype: PropertyKind): Property[]  {
        let classifier: Classifier = Language.getInstance().classifier(typename);
        const foundProperties: Property[] = [];
        if (!!classifier) {
            for (const prop of classifier.properties.values()) {
                if (prop.propertyKind === ptype) {
                    foundProperties.push(prop);
                }
            }
        }
        return foundProperties;
    }

    /**
     * Return the value of `prop' in `element'.
     * For ease of use, always returns a list, even is the property is not a list.
     * @param element
     * @param prop
     */
    public getPropertyValue(element: PiElement, prop: Property): PiElement[] {
        if( prop.isList){
            return element[prop.name];
        } else {
            const value = element[prop.name];
            if( !!value) {
                return [value];
            } else {
                return [];
            }
        }
    }
    createModel(): PiModel {
        return this.pmodel?.constructor();
    }

    createUnit(typeName: string): PiModelUnit {
        return this.units.get(typeName).constructor();
    }

    /**
     * Create a new instance of the class `typeName`.
     * @param typeName
     */
    createConceptOrUnit(typeName: string): PiElement {
        let myType: Concept | ModelUnit = this.concept(typeName);
        if (isNullOrUndefined(myType)) {
            myType = this.unit(typeName);
        }
        return myType?.constructor();
    }


    addModel(model: Model) {
        if(!!this.pmodel) {
            console.error("Language: adding model of type " + model?.typeName + " while there is already a model of type " + this.pmodel.typeName);
        }
        this.pmodel = model;
    }

    addUnit(unit: ModelUnit) {
        this.units.set(unit.typeName, unit);
        unit.subConceptNames = [];
    }

    /**
     * Add a concept definition to this language
     * @param concept
     */
    addConcept(concept: Concept) {
        this.concepts.set(concept.typeName, concept);
    }

    addInterface(intface: Interface) {
        this.interfaces.set(intface.typeName, intface);
    }

    /**
     * Set Language name
     * @param name
     */
    set name(name: string) {
        this.languageName = name;
    }

    get name(): string {
        return this.languageName;
    }

    subConcepts(typeName: string): string[] {
        const concept = this.concept(typeName);
        if (!!concept) {
            return concept.subConceptNames;
        }
        const intface = this.interface(typeName);
        if (!!intface) {
            return intface.subConceptNames;
        }
        return [];
    }

    referenceCreator: (name: string, type: string) => any;

    addReferenceCreator(creator: (name: string, type: string) => any) {
        this.referenceCreator = creator;
    }

    /**
     * Returns true if the piLanguageConcept of 'element', i.e. its metatype,
     * is the same as 'requestedType' or is a subtype of 'requestedType'.
     * @param element
     * @param requestedType
     */
    public metaConformsToType(element: PiElement, requestedType: string): boolean {
        const metatype = element.piLanguageConcept();
        if (metatype === requestedType || Language.getInstance().subConcepts(requestedType).includes(metatype)) {
            return true;
        }
        return false;
    }
}
