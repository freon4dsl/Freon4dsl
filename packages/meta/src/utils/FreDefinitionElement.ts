import { ParseLocation, FreParseLocation } from "./parsingAndChecking/FreGenericParser";

// root of the inheritance structure of all elements that can be present in any of the definitions
export class FreDefinitionElement {
    location: ParseLocation;        // used to store the location information from the PEGJS parser
    agl_location: FreParseLocation;  // used to store the location information from the AGL parser
}
