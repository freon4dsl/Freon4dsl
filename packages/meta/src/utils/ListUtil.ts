export class ListUtil {
    static addIfNotPresent<T>(list: T[], addition: T) {
        if (!!addition && !list.includes(addition)) {
            list.push(addition);
        }
    }

    static addListIfNotPresent<T>(list: T[], additionsList: T[]) {
        additionsList.forEach(extra => {
            ListUtil.addIfNotPresent<T>(list, extra);
        })
    }
}
