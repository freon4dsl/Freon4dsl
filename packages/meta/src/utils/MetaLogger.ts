/**
 * Logging helpers to enable quick filtering of certain log categories.
 */

type MessageFunction = () => string;
type LogMessage = string | MessageFunction;

export class MetaLogger {
    private static muteAll: boolean = false;
    private static muteErrors: boolean = false;
    private static FgRed = "\x1b[31m";
    private static FgBlack = "\x1b[30m";
    private static FgBlue = "\x1b[34m";
    private static FgCyan = "\x1b[36m";
    private static FgMagenta = "\x1b[35m";
    private static FgYellow = "\x1b[33m";

    static muteAllLogs() {
        MetaLogger.muteAll = true;
    }

    static muteAllErrors() {
        MetaLogger.muteErrors = true;
    }

    static unmuteAllLogs() {
        MetaLogger.muteAll = false;
    }

    static filter: string = null;

    static showString(s: string | null) {
        MetaLogger.filter = s;
    }

    category: string;
    active: boolean = true;

    constructor(cat: string) {
        this.category = cat;
    }

    info(msg: LogMessage) {
        if (this.active) {
            this.logToConsole(MetaLogger.FgBlue, this.message(msg));
        }
    }

    warning(msg: LogMessage) {
        if (this.active && !MetaLogger.muteAll) {
            this.logToConsole(MetaLogger.FgYellow, this.message(msg));
        }
    }

    log(msg: LogMessage) {
        if (this.active && !MetaLogger.muteAll) {
            this.logToConsole(MetaLogger.FgBlack, this.message(msg));
        }
    }

    error(msg: LogMessage) {
        if (!MetaLogger.muteErrors) {
            this.logToConsole(MetaLogger.FgRed, "ERROR: " + this.message(msg));
        }
    }

    mute(): MetaLogger {
        this.active = false;
        return this;
    }

    show(): MetaLogger {
        this.active = true;
        return this;
    }

    protected message(msg: LogMessage): string {
        return typeof msg === "string" ? msg : msg();
    }

    protected logToConsole(color: string, message: string): void {
        if (MetaLogger.filter === null) {
            if (this.category.length > 0) {
                console.log(color, this.category + ": " + message, MetaLogger.FgBlack, "");
            } else {
                console.log(color, message, MetaLogger.FgBlack, "");
            }
        } else {
            if (message.includes(MetaLogger.filter)) {
                if (this.category.length > 0) {
                    console.log(color, this.category + ": " + message, MetaLogger.FgBlack, "");
                } else {
                    console.log(color, message, MetaLogger.FgBlack, "");
                }
            }
        }
    }
}
