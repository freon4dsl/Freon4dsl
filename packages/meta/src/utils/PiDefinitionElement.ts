import { ParseLocation, PiParseLocation } from "./parsingAndChecking/PiParser";

// root of the inheritance structure of all elements that can be present in any of the definitions
export class PiDefinitionElement {
    location: ParseLocation;        // used to store the location information from the PEGJS parser
    agl_location: PiParseLocation;  // used to store the location information from the AGL parser
}
