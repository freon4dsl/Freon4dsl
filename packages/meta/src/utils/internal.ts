export * from "./PiDefinitionElement";
export * from "./ListUtil";
export * from "./UserLogger";
export * from "./MetaLogger";
export * from "./parsingAndChecking";
export * from "./generation";
