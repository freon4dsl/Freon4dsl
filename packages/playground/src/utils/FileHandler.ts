import * as fs from "fs";

export class FileHandler {

    /**
     * Reads a string from the file located at 'filepath'. If the
     * file is not present an Error will be thrown.
     * @param filepath
     */
    public stringFromFile(filepath: string): string {
        // read language file
        if (!fs.existsSync(filepath)) {
            console.error(this, "File '" + filepath + "' does not exist, exiting.");
            throw new Error(`File '${filepath}' not found.`);
        }
        return fs.readFileSync(filepath, { encoding: "utf8" });
    }

    /**
     * Writes a string to the file located at 'filepath'. If the
     * file is not present it will be created.
     * May throw an Error if the file cannot be written or created.
     * @param filepath
     * @param output
     */
    public stringToFile(filepath: string, output: string) {
        if (fs.existsSync(filepath)) {
            console.log(this, "FileHandler: file " + filepath + " already exists, overwriting it.");
        }
        fs.writeFileSync(filepath, output);
    }

}
