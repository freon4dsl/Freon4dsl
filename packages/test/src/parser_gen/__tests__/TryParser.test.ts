import { Demo, DemoUnit } from "../language/gen";
import { DemoEnvironment } from "../config/gen/DemoEnvironment";
import { FileHandler } from "../../utils/FileHandler";

describe("Test the parser", () => {
    test( ": read two units and create a model", () => {
        const reader = DemoEnvironment.getInstance().reader;
        const writer = DemoEnvironment.getInstance().writer;
        const fileHandler = new FileHandler();

        let input = fileHandler.stringFromFile("src/parser_gen/__tests__/ParserInput1.txt");
        const unit1: DemoUnit = reader.readFromString(input, "DemoUnit", new Demo()) as DemoUnit;

        unit1.main?.baseEntity.forEach(ent => {
            expect(ent).not.toBeUndefined();
            expect(ent).not.toBeNull();
        });
        unit1.main?.attributes.forEach(attr => {
            expect(attr.declaredType).not.toBeUndefined();
            expect(attr.declaredType).not.toBeNull();
        });

        input = fileHandler.stringFromFile("src/parser_gen/__tests__/ParserInput2.txt");
        const unit2: DemoUnit = reader.readFromString(input, "DemoUnit", new Demo()) as DemoUnit;
        //
        unit2.main?.baseEntity.forEach(ent => {
            expect(ent).not.toBeUndefined();
            expect(ent).not.toBeNull();
        });
        unit2.main?.attributes.forEach(attr => {
            expect(attr.declaredType).not.toBeUndefined();
            expect(attr.declaredType).not.toBeNull();
        });
        //
        const myModels: DemoUnit[] = [];
        myModels.push(unit1);
        myModels.push(unit2);
        const model = Demo.create({ name: "SOME_MODEL", models: myModels });

        for (const unit of model.models) {
            const output = writer.writeToString(unit, 0, false);
            fileHandler.stringToFile(`src/parser_gen/__tests__/Output_${unit.name}.txt`, output);
            expect(output).toMatchSnapshot();
        }
    });
});
