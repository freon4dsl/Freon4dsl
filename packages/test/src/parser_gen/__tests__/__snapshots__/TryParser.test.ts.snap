// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the parser : read two units and create a model 1`] = `
"unit eenNaam {
main: entity1 base anotherEntity,
                   andAnother,
                   andYetAnother
          start
              attr1 : String = \\"myString\\",
              attr2 : Integer = \\"TEKST\\"
          end
unit wide functions:
    DemoFunction functionXXX {
        expression CALL funct555 (  )
        parameters
            DemoVariable variableYYY {
                declaredType entity1
            }
        declaredType String
    }
}"
`;

exports[`Test the parser : read two units and create a model 2`] = `
"unit andereNaam {
main: entity56 base anotherEntity,
                    andAnother,
                    andYetAnother
          start
              attr78 : Boolean = \\"myString\\",
              attr9 : String = \\"TEKST\\"
          end
unit wide functions:
    DemoFunction functionYYY {
        expression CALL funct555 (  )
        parameters
            DemoVariable variableZZZ {
                declaredType entity56
            }
        declaredType Integer
    }
}"
`;
